File: library-service/src/main/kotlin/library/service/api/books/BookResource.kt
package library.service.api.books

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL
import library.service.business.books.domain.BookRecord
import org.springframework.hateoas.ResourceSupport
import org.springframework.hateoas.core.Relation

/** Representation of a [BookRecord] as a REST resource. */
@JsonInclude(NON_NULL)
@Relation(value = "book", collectionRelation = "books")
data class BookResource(
        val isbn: String,
        val title: String,
        val authors: List<String>?,
        val numberOfPages: Int?,
        val borrowed: Borrowed?
) : ResourceSupport()

data class Borrowed(
        val by: String,
        val on: String
)

File: library-ui/src/app/model/book-resource.ts
import {ResourceLink} from "./resource-link";


export class BookResource {
  public isbn: string;
  public title: string;
  public authors: string[];
  public numberOfPages: number;
  public borrowed?: BorrowedResource;
  public _links: BookResourceLinks;
}

class BorrowedResource {
  public by: string;
  public on: Date;
}

class BookResourceLinks {
  public self: ResourceLink;
  public delete?: ResourceLink;
  public borrow?: ResourceLink;
  public return?: ResourceLink;
}
